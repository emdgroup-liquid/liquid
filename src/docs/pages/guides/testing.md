---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

