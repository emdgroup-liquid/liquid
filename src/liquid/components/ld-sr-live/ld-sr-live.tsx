import { Component, h, Listen, State } from "@stencil/core";

/**
 * @virtualProp ref - reference to component
 */
@Component({
  tag: "ld-sr-live",
  styleUrl: "ld-sr-live.shadow.css",
  shadow: true,
})
export class LdSrLive {
  private infoSuffix: string;
  private alertSuffix: string;

  @State() info: string;
  @State() alert: string;

  @Listen("ldSrLiveInfo", {
    target: "window",
    passive: true,
  })
  handleInfo(event: CustomEvent<string>) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.infoSuffix = this.infoSuffix ? "" : "​";
    this.info = event.detail + this.infoSuffix;
  }

  @Listen("ldSrLiveAlert", {
    target: "window",
    passive: true,
  })
  handleAlert(event: CustomEvent<string>) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.alertSuffix = this.alertSuffix ? "" : "​";
    this.alert = event.detail + this.alertSuffix;
  }

  render() {
    return (
      <ld-sr-only>
        <span
          role="status"
          aria-live="polite"
          aria-relevant="all"
          aria-atomic="true"
        >
          {this.info}
        </span>
        <span
          role="alert"
          aria-live="polite"
          aria-relevant="all"
          aria-atomic="true"
        >
          {this.alert}
        </span>
      </ld-sr-only>
    );
  }
}
