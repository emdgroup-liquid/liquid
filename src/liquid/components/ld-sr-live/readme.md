---
eleventyNavigation:
  key: Screen Reader Live
  parent: Components
layout: layout.njk
title: Screen Reader Live
permalink: components/ld-sr-live/
---

# ld-sr-live

Use the `ld-sr-live` component in your application to **conviniently trigger info and alert messages for screen readers**.

---

## How it works

Add the component to your application, preferably close after the opening `<body>` tag. The component is invisible as it uses the [`ld-sr-only` class](components/ld-sr-only/). It listens to two [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent) on the `window`, the `ldSrLiveInfo` and the `ldSrLiveAlert` event. As soon as one of those events reaches the `window`, the component updates the content of one of two contained [`aria-live` reagions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions), either the one with `role="status"` or the one with `role="alert"`, depending on the event. For the content it uses the value of `event.detail`. The "politeness setting" is set to `polite`, meaning that the screen reader delays new messages until it has finished announcing the current ones. Here is an example on how you can trigger an event including an info message that will be spoken by a screen reader:

```js
dispatchEvent(new CustomEvent('ldSrLiveInfo', {
  detail: 'Session expired. You have been logged out.'
}))
```

> **Note**: You should not use more than one `ld-sr-live` component in your app. It acts like a speaker to whome you give a bunch of sentences to speak, and it wouldn't make much sense to have multiple speakers talking at the same time, would it? Technically a screen reader would not speak more than one sentence at a time anyway, no matter how many `aria-live` regions you use. Though you can of course use additional `aria-live` regions, if it makes sense for your application.

## Example

You can test the component in the following example (obviously you'll need to turn on a screen reader of your choice in order to hear the announcements):

{% example %}
<style>
#form-info,
#form-alert {
  display: grid;
  grid-auto-flow: column;
  align-items: flex-end;
  grid-gap: 1rem;
}
</style>

<ld-sr-live></ld-sr-live>

<form id="form-info">
  <ld-label>
    Info message
    <ld-input id="input-info"></ld-input>
  </ld-label>
  <ld-button type="submit">Submit</ld-button>
</form>

<form id="form-alert">
  <ld-label>
    Alert message
    <ld-input id="input-alert"></ld-input>
  </ld-label>
  <ld-button type="submit">Submit</ld-button>
</form>

<script>
const formInfo = document.getElementById('form-info')
const inputInfo = document.getElementById('input-info')
formInfo.addEventListener('submit', ev => {
  ev.preventDefault()
  dispatchEvent(new CustomEvent('ldSrLiveInfo', {
    detail: inputInfo.value || ''
  }))
})

const formAlert = document.getElementById('form-alert')
const inputAlert = document.getElementById('input-alert')
formAlert.addEventListener('submit', ev => {
  ev.preventDefault()
  dispatchEvent(new CustomEvent('ldSrLiveAlert', {
    detail: inputAlert.value || ''
  }))
})
</script>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description            | Type  | Default     |
| -------- | --------- | ---------------------- | ----- | ----------- |
| `ref`    | `ref`     | reference to component | `any` | `undefined` |


----------------------------------------------

 
