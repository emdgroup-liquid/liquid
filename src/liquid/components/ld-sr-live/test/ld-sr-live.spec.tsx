import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdSrLive } from '../ld-sr-live'

describe('ld-sr-live', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdSrLive],
      template: () => <ld-sr-live />,
    })
    expect(page.root).toMatchSnapshot()
  })
})
