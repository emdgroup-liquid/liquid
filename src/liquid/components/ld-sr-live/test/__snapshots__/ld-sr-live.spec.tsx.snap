// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-sr-live renders 1`] = `
<ld-sr-live>
  <mock:shadow-root>
    <ld-sr-only>
      <span aria-atomic="true" aria-live="polite" aria-relevant="text" role="status"></span>
      <span aria-atomic="true" aria-live="polite" aria-relevant="text" role="alert"></span>
    </ld-sr-only>
  </mock:shadow-root>
</ld-sr-live>
`;
